analyzers/src/SonarAnalyzer.CFG/Roslyn/CfgAllPathValidator.cs

```diff
@@ -37,16 +44,32 @@ public abstract class CfgAllPathValidator
         public bool CheckAllPaths() =>
             IsBlockOrAllSuccessorsValid(this.cfg.EntryBlock);
 
-        private bool IsBlockOrAllSuccessorsValid(BasicBlock block) =>
-            !IsInvalid(block) && (IsValid(block) || AreAllSuccessorsValid(block));
+        private bool IsBlockOrAllSuccessorsValid(BasicBlock block)
+        {
+            var isValid = !IsInvalid(block) && (IsValid(block) || AreAllSuccessorsValid(block));
+            lattice[block] = isValid ? BlockResult.TRUE : BlockResult.FALSE;
+            return isValid;
+        }
 
         private bool AreAllSuccessorsValid(BasicBlock block)
         {
-            this.visited.Add(block);
-
-            return !block.SuccessorBlocks.Contains(cfg.ExitBlock)
-                   && block.SuccessorBlocks.Except(visited).Any()
-                   && block.SuccessorBlocks.Except(visited).All(IsBlockOrAllSuccessorsValid);
+            lattice[block] = BlockResult.INCONCLUSIVE;
+            if (!block.SuccessorBlocks.Contains(cfg.ExitBlock) && block.SuccessorBlocks.Any())
+            {
+                foreach (var successorBlock in block.SuccessorBlocks)
+                {
+                    if ((lattice.ContainsKey(successorBlock) && lattice[successorBlock] == BlockResult.FALSE)
```

Comment:
I don't see `TRUE` to be ever read and used. So we might not need that state.

_Thinking out loud:_ What does it actually mean to have `TRUE` there? I see it the same as `INCONCLUSIVE`. So when does the algorithm stop on loops?